# 4.5 - Data Consistency Checks

## This scripts contain the following points:
### 1. Importing Libraries
### 2. Importing Data
### 3. Mixed-Type Data
### 4. Missing Values
### 5. Duplicates
### 6. Task 4.5 
### 7. Exporting Dataframe

# Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Importing Data

In [2]:
# Create path to main folder
path = r'C:\Users\nguye\OneDrive\02-2022 Instacart Basket Analysis'

In [3]:
# Import datasets
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Mixed-Type Data

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
# Checks for mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [7]:
# Change the data type of dataframe
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [8]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
# Create a dataframe to store missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
# Check the shape of dataframe
df_prods.shape

(49693, 5)

In [12]:
# Remove missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
# Check the shape of dataframe
df_prods_clean.shape

(49677, 5)

# Duplicates

In [14]:
# Search for duplicates in dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [15]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
# Check shape of dataframe
df_prods_clean.shape

(49677, 5)

In [17]:
# Create new dataframe with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
# Verify that duplicates were removed
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5 - Data Consistency Checks

## Question 2

In [19]:
# Check descriptive statistics for df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### One issue to note is the "unnamed" column. This column gives no meaningful data and will be removed. It was added as an index during one of the previous steps. All other columns show no inconsistencies. 

In [20]:
# Remove Unnamed column
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

## Question 3

In [21]:
# Check for mixed-typed data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

### No mixed-type data was found in the dataframe

## Question 5

In [22]:
# Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [23]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [24]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [25]:
df_ords_nan.describe()

order_id        user_id  order_number  orders_day_of_week  \
count  2.062090e+05  206209.000000      206209.0       206209.000000   
mean   1.708462e+06  103105.000000           1.0            2.754118   
std    9.881299e+05   59527.555167           0.0            2.076205   
min    2.000000e+01       1.000000           1.0            0.000000   
25%    8.507300e+05   51553.000000           1.0            1.000000   
50%    1.706246e+06  103105.000000           1.0            3.000000   
75%    2.564292e+06  154657.000000           1.0            5.000000   
max    3.421081e+06  206209.000000           1.0            6.000000   

       order_hour_of_day  days_since_prior_order  
count      206209.000000                     0.0  
mean           13.626597                     NaN  
std             4.223769                     NaN  
min             0.000000                     NaN  
25%            11.000000                     NaN  
50%            14.000000                     NaN  
75%            17.000000                     NaN  
max            23.000000                     NaN

### The column "days_since_prior_order" contains 206,209 NaN values. This is attributed to the fact that this is the user's only purchase and therefore there can be no value for "days_since_prior_order.

## Question 6

### These missing values are significant in the fact that there can be no logical value inputed instead, as these are the user's first purchase. Once they make another purchase, than the values will change, but in this case, the missing values will be left as is. 

## Question 7

In [26]:
# Find duplicates in df_ords
df_ords_dups = df_ords_nan[df_ords_nan.duplicated()]

In [27]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### No duplicates were found in the dataframe

## Question 8

### As no duplicates were found in the dataframe, no action was necessary

# Exporting Dataframes

In [30]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))